daemon/k8s_watcher.go

```diff
@@ -1735,127 +1708,109 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
-		} else if ruleCopyParseErr != nil {
-			// This handles the case where the initial instance of this
-			// rule was imported into the policy repository successfully
-			// (policyImportErr == nil), but, the rule has been updated
-			// in the store soon after, and is now invalid. As such,
-			// the rule is not OK because it cannot be imported due
-			// to parsing errors, and cannot be enforced because it is
-			// not OK.
-			cnpUpdateErr = updateCNPNodeStatus(serverRuleCpy, false, false, ruleCopyParseErr, rev, serverRuleCpy.Annotations)
-		} else {
-			// If the deadline by the above context, then not all
-			// endpoints are enforcing the given policy, and
-			// waitForEpsErr will be non-nil.
-			cnpUpdateErr = updateCNPNodeStatus(serverRuleCpy, waitForEPsErr == nil, true, waitForEPsErr, rev, serverRuleCpy.Annotations)
+			cnpError = policyImportErr
 		}
 
+		updatedCNP, cnpUpdateErr = updateCNPNodeStatus(cnp, overallErr == nil, policyImportErr == nil, cnpError, rev)
 		if cnpUpdateErr == nil {
-			logger.WithField("status", serverRuleCpy.Status).Debug("successfully updated with status")
-			break
+			logger.WithField("status", updatedCNP.Status).Debug("successfully updated with status")
+			return overallErr
 		}
 
 		// Wait a small amount of time to try to update again.
 		logger.WithError(cnpUpdateErr).Warningf("update of CNP failed; sleeping for %s amount of time before trying again", updateWaitDuration)
 		time.Sleep(updateWaitDuration)
 	}
 
-	if cnpUpdateErr != nil {
-		return cnpUpdateErr
-	} else {
-		return waitForEPsErr
+	// If the process of waiting for endpoints has been successful but the
+	// CNP status report failed, report an error to cause the controller to
+	// retry.
+	if overallErr != nil && cnpUpdateErr != nil {
+		overallErr = cnpUpdateErr
 	}
+
+	return overallErr
 }
 
-func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool, err error, rev uint64, annotations map[string]string) error {
-	var (
-		cnpns cilium_v2.CiliumNetworkPolicyNodeStatus
-		err2  error
-	)
+func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool, importError error, rev uint64) (*cilium_v2.CiliumNetworkPolicy, error) {
+	ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
 
+	// Retrieve the latest revision from the server to minimize the risk of
+	// a locally corrupted CNP makes its way back into the apiserver
+	latestServerRev, err := ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Get(cnp.GetName(), metav1.GetOptions{})
 	if err != nil {
-		cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-			Enforcing:   enforcing,
-			Error:       err.Error(),
-			OK:          ok,
-			LastUpdated: cilium_v2.NewTimestamp(),
-			Annotations: annotations,
-		}
+		return nil, err
+	}
+
+	// Create a copy so it can be modified to update the status section
+	latestServerCopy := latestServerRev.DeepCopy()
```

Comment:
We were doing the `DeepCopy()` to prevent corruption in the cache, the same way we are doing it here, I really don't understand what we can possible gain while making it slower by fetching it from kube-apiserver every time we want to update the status.

Reply:
While this is already merged, I agree with André. I don't see the added benefits here.

Reply:
The DeepCopy here is unchanged from previous code and there is no claim that moving the DeepCopy here makes a difference.

The benefit is not depending on a shared cache to which all users can write to. A single mistake in modifying a CNP in the store in any of the agent code can corrupt the CNP for the entire cluster. That's a tremendous unnecessary risk.

It's a simple principle of distributed systems to limit modification of shared resources. The easier and ressilient that code, the better. This is not about what is better if code is bug free, this is about what is more resilient if things go wrong.